base/math.jl

```diff
@@ -392,7 +392,7 @@ julia> exp10(2)
 """
 exp10(x::AbstractFloat) = 10^x
 
-for f in (:sinh, :cosh, :tanh, :atan, :asinh, :exp, :expm1)
+for f in (:sin, :cos, :tan, :sinh, :cosh, :tanh, :atan, :acos, :asin, :asinh, :acosh, :atanh, :exp, :expm1, :log)
     @eval ($f)(x::AbstractFloat) = error("not implemented for ", typeof(x))
 end
```

Comment:
It should be a `MethodError`, but we should also move the conversion that causes the StackOverflow error to the same place, so that we don't miss methods.

```suggestion
for f in (:sin, :cos, :tan, :sinh, :cosh, :tanh, :atan, :acos, :asin, :asinh, :acosh, :atanh, :exp, :expm1, :log)
    @eval begin
        ($f)(x::Real) = ($f)(float(x))
        ($f)(x::AbstractFloat) = throw(MethodError($f, (x,)))
    end
end
```

Reply:
Or perhaps a better solution, which would help with issues like `Dual`s from ForwardDiff (which aren't caught by the current `AbstractFloat` dispatch):
```suggestion
for f in (:sin, :cos, :tan, :sinh, :cosh, :tanh, :atan, :acos, :asin, :asinh, :acosh, :atanh, :exp, :expm1, :log)
    @eval function ($f)(x::Real)
        xf = float(x)
        x === xf && throw(MethodError(f, (x,)))
        return ($f)(xf)
    end
end
```

Reply:
Would this version catch `AbstractFloat`s which are not `Real`s?

Reply:
This cannot happen.
```julia
julia> AbstractFloat <: Real
true
```

Reply:
OK, then this is fine

Reply:
@dkarrasch apparently it can !

Reply:
It can?

Reply:
I think the issue is that we need to remove the existing definitions for e.g. `expm1(::AbstractFloat)`.